File: Matlab/kolus_initDAQ.m
function kolus_initDAQ()
% Initialize data acquisition setup using National Instruments Card
K_config;
global S
S.Daq = daq.createSession('ni');
tag.type_Out = {};

for i  = 1:length(daq_channels)
    Ch_current = daq_channels{i};
    switch Ch_current{3}
        case 'digital output'
            io_type = 'OutputOnly';
            tag.type_Ch(i) = 0;
            tag.type_Out = [tag.type_Out Ch_current{end}];
            S.Ch(i) = S.Daq.addDigitalChannel(daq_dev,Ch_current{2}, io_type);
        case 'digital input'
            io_type = 'InputOnly';
            tag.type_Ch(i) = 1;
            S.Ch(i) = S.Daq.addDigitalChannel(daq_dev,Ch_current{2}, io_type);
        case 'analog input'
            io_type = 'Voltage';
            tag.type_Ch(i) = 1;
            S.Ch(i) = S.Daq.addAnalogInputChannel(daq_dev,Ch_current{2}, io_type);
            S.Ch(i).Range = Ch_current{5};
            S.Ch(i).TerminalConfig = daq_terminal;
        case 'analog output'
            io_type = 'Voltage';
            tag.type_Ch(i) = 0;
            tag.type_Out = [tag.type_Out Ch_current{end}];
            S.Ch(i) = S.Daq.addAnalogOutputChannel(daq_dev,Ch_current{2}, io_type);
            S.Ch(i).Range = Ch_current{5};
        otherwise
            warning('Error in K_config input/output field')
            pause(1e3)
    end
    S.Ch(i).Name = Ch_current{1};
end

S.Daq.Rate = max(tag.Rates);

tag.fs = S.Daq.Rate; %NI Rate - all recording happens at max rate
tag.folder = cd; %experiment folder
guidata(gcf, tag) %pass tag on for modifications/saving

%Pre-allocate buffer for saving
SaveData = nan(sum(tag.refresh_time*tag.Rates), 1);

addlistener(S.Daq,'DataAvailable',...
    @(src,obj)Kolus_datasave(src,obj,SaveData, tag.Rates));
% addlistener(S.Daq, 'DataRequired', @Kolus_continuous); %for continuous


function Kolus_datasave(src,obj, SaveData, Rates)

global DAQ_buffer
Fs = src.Rate;
DAQ_buffer.Time = obj.TriggerTime;

if ~isempty(DAQ_buffer.Fid)
%     tic
    chunk_end = 0;
    for i = 1:length(Rates)
        chunk_start = chunk_end + 1;
        chunk_end = (chunk_start - 1) + ( size(obj.Data,1) * (Rates(i) / Fs) );
        SaveData(chunk_start:ceil(chunk_end)) = ...
            obj.Data(1:(Fs/Rates(i)):end, i);
    end
    fwrite(DAQ_buffer.Fid, SaveData, 'double'); %avg. time .2 MS for every 100ms data
    start_b = DAQ_buffer.Index;
    DAQ_buffer.Buffer(start_b:(length(SaveData)+ start_b - 1)) = SaveData;
    DAQ_buffer.Index = start_b + length(SaveData);
%     toc
end

function Kolus_continuous(src,~)
%%%Queue 10 seconds of zeros - no stimulus
queueOutputData(src, zeros(src, 1));

File: Python/kolus_initDAQ.py
import pyvisa
import numpy as np


def kolus_initDAQ():
    global S  # Avoid global variables if possible

    load_config()  # Assuming your K_config setup works similarly 
    tag.type_Out = []  # Empty list
    S = {}  # A dictionary instead of a struct might be cleaner

    rm = pyvisa.ResourceManager()
    daq = rm.open_resource(daq_dev)  # Replace with your specific connection method

    for Ch_current in daq_channels:
        io_type = ''  
        tag_type = -1  # Initialize in case of error

        if Ch_current[2] == 'digital output':
            io_type = 'DO'  # Assuming PyVISA uses short codes
            tag_type = 0
            tag.type_Out.append(Ch_current[-1])
        elif Ch_current[2] == 'digital input':
            io_type = 'DI' 
            tag_type = 1
        # ... Other cases similarly ... 

        else:
            print("Warning: Error in K_config input/output field")  # Replace with logging
            continue  # Or raise an exception

        # Assuming PyVISA has similar functionalities:
        S[Ch_current[1]] = daq.write(f"CONFIG:CHANNEL {Ch_current[1]},{io_type}")  # Placeholder command
        tag['type_Ch'].append(tag_type) 

    tag['fs'] = max(tag['Rates'])    
    tag['folder'] = os.getcwd()  
    # ... GUI handling will be Python specific ....

    # Allocate buffer and set listener
    SaveData = np.zeros(int(sum(tag['refresh_time'] * tag['Rates'])))
    daq.events.register_data_available_event(kolus_datasave, SaveData, tag['Rates'])

def kolus_datasave(instrument, data, rates):  
    """
    This function retrieves data from the NI device using PyVISA and processes it 
    based on the provided rates.

    Args:
        instrument: PyVISA instrument object connected to the NI device
        data: 
        rates: Sampling rates for each channel

    Returns:
        None (data is assumed to be saved elsewhere)
    """

    # Assuming binary data with format: <number of samples>,<channel 1 data (16-bit signed integers)>, ...
    # Replace with the actual format if different
    data_format_string = "<I{ch}"

    # Get number of channels from configured rates
    num_channels = len(rates)
    data_format_string = data_format_string.format(ch=num_channels * "h")  # 'h' for short (16-bit) signed integers

    try:
        # Read binary data from the instrument
        raw_data = instrument.read_raw(data_format_string.encode())

        # Assuming data starts from the second element (after number of samples)
        processed_data = np.frombuffer(raw_data[1:], dtype=np.int16).reshape(-1, num_channels)  

        # Separate data for each channel based on sampling rates
        channel_data = []
        start = 0
        for rate in rates:
            end = start + int(len(processed_data) / rate)
            channel_data.append(processed_data[start:end])
            start = end

        # ... Save the 'channel_data' list (implementation depends on your saving logic) ...

    except Exception as e:
        print(f"Error reading data from NI device: {e}")

